client/state/add-reducer.js

```diff
@@ -3,24 +3,41 @@
 /**
  * Internal Dependencies
  */
-
 import { APPLY_STORED_STATE } from 'state/action-types';
 import { getStateFromLocalStorage } from 'state/initial-state';
 
+const initializations = {};
+
+function normalizeKey( key ) {
+	return Array.isArray( key ) ? key.join( '.' ) : key;
+}
+
+async function initializeState( store, reducer, storageKey ) {
+	const storedState = await getStateFromLocalStorage( reducer, storageKey );
+
+	if ( storedState ) {
+		store.dispatch( { type: APPLY_STORED_STATE, storageKey, storedState } );
+	}
+}
+
 // For a given store, creates a function that adds a new reducer to the store,
 // and loads (asynchronously) and applies the persisted state for it.
 export const addReducerToStore = store => async ( key, reducer ) => {
-	store.addReducer( key, reducer );
-
 	const { storageKey } = reducer;
-	if ( ! storageKey ) {
-		return;
-	}
+	const normalizedKey = normalizeKey( key );
+	let init = initializations[ normalizedKey ];
 
-	const storedState = await getStateFromLocalStorage( reducer, storageKey );
-	if ( ! storedState ) {
-		return;
+	if ( ! init ) {
+		store.addReducer( key, reducer );
+
+		if ( storageKey ) {
+			init = initializeState( store, reducer, storageKey );
+		} else {
+			init = Promise.resolve();
+		}
+
+		initializations[ normalizedKey ] = init;
 	}
 
-	store.dispatch( { type: APPLY_STORED_STATE, storageKey, storedState } );
+	await init;
```

Comment:
nit: the function doesn't really need to async: just returning the `init` promise here has the same behavior. There would be some gotchas if there were try/catch blocks, but there are none.

Reply:
Very good point! I'll make that change.

Reply:
Done.